### ISSUE TITLE
Redundant vtable entries are generated, resulting in unnecessary amounts of LLVM IR being emitted

### ISSUE THREAD

**Issue by zesterer** [rust-lang/rust#94975]
Rust appears to generate LLVM IR for vtable entries unnecessarily, such as for types that never get coerced into a trait object. This realisation came as a result of discussion [here](https://github.com/zesterer/chumsky/issues/13).

I've also theorised that it does this for all trait methods outright, although I've yet to confirm that this is the case. It might be that in this latter case, the compiler only generates trait methods for types that it *observes* that implement a given trait, although this still results in *a lot* of redundant LLVM IR being generated, but might be less easy to fix owing to issues with compilation units.

The result of this is that the compilation performance of combinator-like APIs is massively pessimised, with LLVM needing to eat through enormously quantities of redundant IR, only for it to be dropped by the linker.

For example, the following code, when compiled with `cargo rustc -- --emit=llvm-ir`, results in Rust generating vtable entries of `Foo` for both `u8` and `u16`, despite the fact that only `u8` is ever actually coerced into a trait object.

```rust
// Unimportant, just makes the LLVM IR easier to read
#![feature(lang_items, start, bench_black_box)]
#![no_std]
use core::panic::PanicInfo;
use core::hint::black_box;

// This is the important bit

trait Foo {
    fn x(&self);
}
impl Foo for u8 {
    fn x(&self) {
        unsafe { use_foo("xu8"); }
    }
}
impl Foo for u16 {
    fn x(&self) {
        unsafe { use_foo("xu16"); }
    }
}

#[start]
fn main(_argc: isize, _argv: *const *const u8) -> isize {
    let foo: &dyn Foo = &123_u8;
    foo.x();
    123_u8.x();
    123_u16.x();
    0
}

// This only exists to force Rust to definitely generate code and is not relevant to the issue at hand
#[no_mangle] pub extern "C" fn use_foo(x: &'static str) { let _ = black_box(x); }
#[panic_handler] fn panic(_info: &PanicInfo) -> ! { loop {} }
#[lang = "eh_personality"] extern "C" fn eh_personality() {}
```

Initially, I thought that Rust might have been conservatively generating vtable entries on the assumption that other compilation units might be making use of those vtable entries, but in hindsight this doesn't make sense because each compilation unit gets a copy of any vtable entries that it requires *anyway* and so there should be no reason for the compiler to need to generate vtable entries for types that it does not observe being coerced into a trait object.

### Meta

This appears to occur for all compiler versions, including the latest nightly.

### I don't know what I'm doing

I've not done serious work on the Rust compiler myself, so it might be that I'm missing some essential piece of the puzzle. Apologies if I've made errors in my account above.

**Comment by erikdesjardins** [comment]
When compiling that example locally or on godbolt: https://godbolt.org/z/9Ecch96jG, only one vtable (`@vtable.0`) is generated. Perhaps it's reduced too much from the the original issue?

(Aside: I would be a little surprised if the issue you're seeing is caused by vtables specifically, since they should be generated on demand when we codegen an unsizing coercion, e.g. [here](https://github.com/rust-lang/rust/blob/52b34550aca5f7dd7e152f773e3ab786acb86f6f/compiler/rustc_codegen_ssa/src/base.rs#L176-L183) and [here](https://github.com/rust-lang/rust/blob/52b34550aca5f7dd7e152f773e3ab786acb86f6f/compiler/rustc_codegen_ssa/src/meth.rs#L61). But I don't want to discourage investigating this, there could be something else going wrong.)

**Comment by zesterer** [comment]
Perhaps I'm reading the LLVM IR incorrectly and misindentifying a normal method call.

That said, there does seem to be *something* generating superfluous LLVM IR in combinator-heavy code (see [here](https://github.com/fasterthanlime/netherquote)), and it seems to result in significantly less LLVM IR being emitted when moving dynamic uses of a trait to a helper trait with a blanket impl [like this](https://github.com/zesterer/chumsky/commit/f6f39da671ea2807c922fb6db3398452c96d6dcd).

I'm going to do a little more testing today in an attempt to come up with a definitive example.

**Comment by erikdesjardins** [comment]
When compiling https://github.com/fasterthanlime/netherquote locally with `cargo llvm-lines` (optionally with `--release`), with `chumsky` patched to a local clone:
```
# in Cargo.toml
[patch.crates-io]
chumsky = { path = '../chumsky' }
```

I see minimal difference in LLVM IR lines (<100) between:
- https://github.com/zesterer/chumsky/commit/f6f39da671ea2807c922fb6db3398452c96d6dcd (using `dyn DynParser`) [0] [1]
- https://github.com/erikdesjardins/chumsky/commit/373752c7baa7ee62f0a0dbf011e93afc837d6255 (removing `DynParser` and using `dyn Parser` directly) [2] [3].

This is basically what I would expect. Both `dyn DynParser` and `dyn Parser` should generate vtables for the same set of types, and the same two methods (`parse_inner_verbose` and `parse_inner_silent`), since most methods in `Parser` are `Self: Sized`.

The significant part of [f6f39da](https://github.com/zesterer/chumsky/commit/f6f39da671ea2807c922fb6db3398452c96d6dcd) was the use of `invoke_unsized`. Reverting that part does cause a significant increase in LLVM IR lines (~20k) [4] [5]. This is probably for the usual reasons: using dynamic dispatch avoids monomorphizing a bunch of copies of generic functions for each unique parser implementation.

[0] [dyn_dynparser.txt](https://github.com/rust-lang/rust/files/8281323/dyn_dynparser.txt)
[1] [dyn_dynparser_release.txt](https://github.com/rust-lang/rust/files/8281324/dyn_dynparser_release.txt)
[2] [dyn_just_parser.txt](https://github.com/rust-lang/rust/files/8281325/dyn_just_parser.txt)
[3] [dyn_just_parser_release.txt](https://github.com/rust-lang/rust/files/8281327/dyn_just_parser_release.txt)
[4] [no_invoke_unsized.txt](https://github.com/rust-lang/rust/files/8281330/no_invoke_unsized.txt)
[5] [no_invoke_unsized_release.txt](https://github.com/rust-lang/rust/files/8281335/no_invoke_unsized_release.txt)

**Comment by dotdash** [comment]
@erikdesjardins did you compile the OP's example or the one you shared on godbolt?

What happens is that godbolt compiles the crate as a library. Note that it's `pub fn main` there, without the `pub`, the `main` function would be omitted and there would be no vtables at all.

When compiling the original example as a binary crate, there are actually two vtables, but they're not both related to `Foo`, one is passed to `std::rt::lang_start_internal` which expects the actual `main` as a `dyn Fn`.

Adding more implementations of `Foo` without using them doesn't change the amount of generated LLVM IR. The second vtable was just a red herring.

@zesterer In the chumsky issue you had [this playground link](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=4300282e043779f476bc4ec56102fd86) - this also compiles the code as a library when you use "SHOW LLVM IR", and there are actually no vtables in the generated code at all. Did you by chance just look for the functions being defined in the code? That doesn't mean that there are any vtables, they would show up as `@vtable.0`, `@vtable.1` etc. in the IR. That the functions are being generated although they seem to be unused is due to them being public and the crate being compiled as a library. Other crates may use them.

**Comment by zesterer** [comment]
@dotdash I think my initial suspicion of LLVM IR output might have been wrong.

That said, there is definitely *something* with exponential complexity going on in the compiler that shouldn't have exponential complexity (incidentally, my own language has a type system with typeclasses very similar to Rust, and code of similar complexity to that [seen here](https://github.com/fasterthanlime/netherquote) compiles significantly faster (the difference is exponential, not linear).)

However, I don't have the domain-specific knowledge of `rustc` required to determine what exactly it is that causes the compiler to choke on these kinds of programs.

**Comment by erikdesjardins** [comment]
@dotdash yeah I should've been clearer in my first comment--using a fake `pub fn main` in a library crate makes the IR easier to read by getting rid of the (usually irrelevant) startup code.

Everything mentioned in my second comment was compiled as a binary crate. (That is, assuming `cargo llvm-lines` doesn't do anything weird)
